### PR TITLE
Replacing synchronized blocks with ReentrantLocks in core part of SDK

### PR DESCRIPTION
**Pull request by Myllyenko**
Continuation of https://github.com/ydb-platform/ydb-java-sdk/pull/338

### CODE REVIEW

File: core/src/main/java/tech/ydb/core/impl/call/ReadWriteStreamCall.java [hunk 1]
```diff
@@ -126,40 +137,53 @@ public void onMessage(R message) {
...
+        } finally {
+            callLock.unlock();
         }
     }
 
     @Override
     public void onClose(io.grpc.Status status, @Nullable Metadata trailers) {
         if (logger.isTraceEnabled()) {
-            logger.trace("ReadWriteStreamCall[{}] closed with status {}", status);
+            logger.trace("ReadWriteStreamCall[{}] closed with status {}", traceId, status);
```
Comment (Myllyenko) [inline core/src/main/java/tech/ydb/core/impl/call/ReadWriteStreamCall.java:186]: First log argument was absent

File: core/src/main/java/tech/ydb/core/impl/call/ReadWriteStreamCall.java [hunk 2]
```diff
@@ -28,10 +28,11 @@
  * @param <W> type of message to be sent to the server
  */
 public class ReadWriteStreamCall<R, W> extends ClientCall.Listener<R> implements GrpcReadWriteStream<R, W> {
-    private static final Logger logger = LoggerFactory.getLogger(GrpcTransport.class);
+    private static final Logger logger = LoggerFactory.getLogger(ReadWriteStreamCall.class);
```
Comment (Myllyenko) [inline core/src/main/java/tech/ydb/core/impl/call/ReadWriteStreamCall.java:31]: Looks like a copy-paste error in a logger definition

Comment (pnv1) [reply core/src/main/java/tech/ydb/core/impl/call/ReadWriteStreamCall.java:31]: Could sometimes be intended to setup logging level and/or filter logs by a single name (like "GrpcTransport") to see whats happening on this level. Not sure about this particular case

**Review by pnv1 — APPROVED** [review]

File: core/src/main/java/tech/ydb/core/impl/call/ReadWriteStreamCall.java (hunk 2 above)
Comment (alex268) [reply core/src/main/java/tech/ydb/core/impl/call/ReadWriteStreamCall.java:31]: Yes, it is not a copy-paste error, is was made for filtering logs by one class name

**Comment by alex268** [comment]
I am not sure that we have to remove `synchronized` from call classes. They are just simple wrappers on GrpcCall and all their calls are not blocking. In this case, `synchronized` just adds thread safety to GrpcCall, especially with case data race on start() and cancel(). Replacing them by ReentrantLock makes them more complex and may reduce a performance

File: core/src/main/java/tech/ydb/core/impl/call/ReadWriteStreamCall.java (hunk 2 above)
Comment (Myllyenko) [reply core/src/main/java/tech/ydb/core/impl/call/ReadWriteStreamCall.java:31]: Restored and commented previous value

**Comment by Myllyenko** [comment]
> I am not sure that we have to remove `synchronized` from call classes. They are just simple wrappers on GrpcCall and all their calls are not blocking. In this case, `synchronized` just adds thread safety to GrpcCall, especially with case data race on start() and cancel(). Replacing them by ReentrantLock makes them more complex and may reduce a performance

Reverted changes in `StreamCall`s.

**Review by alex268 — APPROVED** [review]

**Pull request merged**